### PR TITLE
fix for closed profiles/guides. 

### PR DESCRIPTION
**Pull request by Shkolik**
Code should decide by itself what result surface to create. Leave it to the api user leads to unstable/unpredictable results and, without harsh modifications of code, not practical.

The main problem was that when profiles are closed api was expecting additional guide from user - copy of the first guide at the end of collection. It works, but only when 1st guide in a collection intersect profile at lowest parameter, so basically collection already sorted. All other cases were leading to failure, since before sorting we do not know what guide will be the first one and as result cannot decide witch guide to duplicate. 

It's just a draft that allowed me to play a bit more with a lib and finally build surface over one of the test data example (nacelle). Hope it will help improve the library that has pretty good potential. 

If you have any suggestions - I'll be happy to discuss. Meanwhile I'll continue my journey over the code trying to make it fork for my application.

image

### CODE REVIEW

**Comment by rainman110** [comment]
Dear @Shkolik . Thank you for your first contribution. I really welcome an addition, that one does not *need* to provide a duplicate curve / profile in case of a closed surface.

Currently though, most of the gordon surface unit tests are failing, see

```
C:\src\occ_gordon\build\build-occt-762\tests\unittests>Release\occ_gordon-unittests.exe --gtest_filter=BSplineAlgorithms/GordonSurface*
Note: Google Test filter = BSplineAlgorithms/GordonSurface*
[==========] Running 24 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 24 tests from BSplineAlgorithms/GordonSurface
[ RUN      ] BSplineAlgorithms/GordonSurface.testFromBRep/0
[       OK ] BSplineAlgorithms/GordonSurface.testFromBRep/0 (21 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testFromBRep/1
unknown file: error: C++ exception with description "Invalid index in BSplineApproxInterp::InterpolatePoint" thrown in the test body.

[  FAILED  ] BSplineAlgorithms/GordonSurface.testFromBRep/1, where GetParam() = "full_nacelle" (10 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testFromBRep/2
[       OK ] BSplineAlgorithms/GordonSurface.testFromBRep/2 (78 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testFromBRep/3
unknown file: error: C++ exception with description "Invalid index in BSplineApproxInterp::InterpolatePoint" thrown in the test body.

[  FAILED  ] BSplineAlgorithms/GordonSurface.testFromBRep/3, where GetParam() = "spiralwing" (8 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testFromBRep/4
[       OK ] BSplineAlgorithms/GordonSurface.testFromBRep/4 (10 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testFromBRep/5
[       OK ] BSplineAlgorithms/GordonSurface.testFromBRep/5 (10 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testFromBRep/6
unknown file: error: C++ exception with description "Invalid index in BSplineApproxInterp::InterpolatePoint" thrown in the test body.

[  FAILED  ] BSplineAlgorithms/GordonSurface.testFromBRep/6, where GetParam() = "wing3" (4 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testFromBRep/7
[       OK ] BSplineAlgorithms/GordonSurface.testFromBRep/7 (49 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testFromBRep/8
unknown file: error: C++ exception with description "Invalid index in BSplineApproxInterp::InterpolatePoint" thrown in the test body.

[  FAILED  ] BSplineAlgorithms/GordonSurface.testFromBRep/8, where GetParam() = "helibody" (103 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testFromBRep/9
unknown file: error: C++ exception with description "Invalid index in BSplineApproxInterp::InterpolatePoint" thrown in the test body.

[  FAILED  ] BSplineAlgorithms/GordonSurface.testFromBRep/9, where GetParam() = "fuselage1" (26 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testFromBRep/10
unknown file: error: C++ exception with description "Invalid index in BSplineApproxInterp::InterpolatePoint" thrown in the test body.

[  FAILED  ] BSplineAlgorithms/GordonSurface.testFromBRep/10, where GetParam() = "fuselage2" (8 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testFromBRep/11
[       OK ] BSplineAlgorithms/GordonSurface.testFromBRep/11 (264 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/0
[       OK ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/0 (3 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/1
[       OK ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/1 (12 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/2
[       OK ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/2 (8 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/3
[       OK ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/3 (8 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/4
[       OK ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/4 (4 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/5
[       OK ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/5 (2 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/6
[       OK ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/6 (4 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/7
[       OK ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/7 (3 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/8
[       OK ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/8 (135 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/9
[       OK ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/9 (29 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/10
[       OK ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/10 (12 ms)
[ RUN      ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/11
[       OK ] BSplineAlgorithms/GordonSurface.testIntersectionRegressions/11 (23 ms)
[----------] 24 tests from BSplineAlgorithms/GordonSurface (871 ms total)

[----------] Global test environment tear-down
[==========] 24 tests from 1 test suite ran. (872 ms total)
[  PASSED  ] 18 tests.
[  FAILED  ] 6 tests, listed below:
[  FAILED  ] BSplineAlgorithms/GordonSurface.testFromBRep/1, where GetParam() = "full_nacelle"
[  FAILED  ] BSplineAlgorithms/GordonSurface.testFromBRep/3, where GetParam() = "spiralwing"
[  FAILED  ] BSplineAlgorithms/GordonSurface.testFromBRep/6, where GetParam() = "wing3"
[  FAILED  ] BSplineAlgorithms/GordonSurface.testFromBRep/8, where GetParam() = "helibody"
[  FAILED  ] BSplineAlgorithms/GordonSurface.testFromBRep/9, where GetParam() = "fuselage1"
[  FAILED  ] BSplineAlgorithms/GordonSurface.testFromBRep/10, where GetParam() = "fuselage2"
```

Please make sure, that the tests are succeeding first.

Also, it would be nice if you could add a test, where the duplicate curve is removed and the order of curves in randomized. This should check, that the algorithm figures out, what curve it has to replicate.

**Comment by rainman110** [comment]
I think in all these tests, on surface direction is closed, so the one curve is duplicate. I find it important, that the algorithm can handle also the case where the curves are duplicate as other software relies on it. so you basically need to figure out, whether to add a curve or not.

**Comment by Shkolik** [comment]
> I think in all these tests, on surface direction is closed, so the one curve is duplicate. I find it important, that the algorithm can handle also the case where the curves are duplicate as other software relies on it. so you basically need to figure out, whether to add a curve or not.

Yep, you are right. I'll make a fix for that.

**Comment by Shkolik** [comment]
I pushed fix that should solve issue with failed tests. Unfortunately I cannot run tests on my machine with current OCC installation. Could you please provide command you are using to build tests? Maybe I'm missing something.

**Comment by rainman110** [comment]
The tests are running through. The results are also looking good. Well done! I am looking at the code tomorrow, need to sleep ;)

Here are the commands to run the tests by yourself.

```
# enable unit tests
cmake -DOCC_GORDON_BUILD_TESTS=ON .

# building
cmake --build .

cd tests/unittests
./occ_gordon-unittests

cd ../apitests
./occ_gordon-apitest
```

It uses the google test framework. You can run specific tests via wildcards e.g.

    ./occ_gordon-unittests --gtest_filter=BSplineAlgorithms/GordonSurface.*

**Comment by Shkolik** [comment]
Another suggestion. Right now API interpolate_curve_network expect std::vector<Handle(Geom_Curve) as profiles and guides. It would be better to change it to  std::vector<Handle(Geom_BSplineCurve) by 2 reasons:
1) We are aiming for B-Spline surface as a result -> need B-Splines anyway
2) conversion from  Geom_Curve to Geom_BSplineCurve right now done inside a constructor anyway and not the best way. Sometimes GeomConvert::CurveToBSplineCurve may fail, but there is a way to use ShapeConstruct_Curve or even just Handle(Geom_BSplineCurve)::DownCast(curve) if initial curve is already B-Spline. 
3) It's impossible to transform Geom_Curve but easy with Geom_BSplineCurve. For real world scenario, when user extracting geometry entity from topology (in this example from Edge) before feeding algorithm geometry should be transformed to real space coordinates or recreated in this coordinates. Transform is much easier.

Doing so will ease use of library for the end user and it's not a big change. Maybe I could make 2 versions of  occ_gordon::interpolate_curve_network - one for Geom_BSplineCurve and another as it's right now. 

What do you think?

**Comment by rainman110** [comment]
> Doing so will ease use of library for the end user and it's not a big change. Maybe I could make 2 versions of occ_gordon::interpolate_curve_network - one for Geom_BSplineCurve and another as it's right now.

You could add a second function (ideally using overloads) with Geom_BSplineCurve arguments. The "old" function then should convert the Geom_Curves and call the B-Spline-based function.

File: src/internal/InterpolateCurveNetwork.cpp [hunk 1]
```diff
@@ -34,9 +34,9 @@ T Clamp(T val, T min, T max)
     return std::max(min, std::min(val, max));
 }
 
-InterpolateCurveNetwork::InterpolateCurveNetwork(const std::vector<Handle (Geom_Curve)> &profiles,
-                                                           const std::vector<Handle (Geom_Curve)> &guides,
-                                                           double spatialTol)
+InterpolateCurveNetwork::InterpolateCurveNetwork(const std::vector<Handle(Geom_Curve)>& profiles,
+    const std::vector<Handle(Geom_Curve)>& guides,
+    double spatialTol)
```
Comment (rainman110) [inline src/internal/InterpolateCurveNetwork.cpp:39]: Just a remark. Please don't commit re-formatted code. It makes it hard to find out the true code differences.

File: src/internal/InterpolateCurveNetwork.cpp [hunk 2]
```diff
@@ -45,90 +45,109 @@ InterpolateCurveNetwork::InterpolateCurveNetwork(const std::vector<Handle (Geom_
...
+        for (const auto& uniqueProfile : uniqueProfiles) {
+            if (profile->Value(profile->FirstParameter()).IsEqual(uniqueProfile->Value(uniqueProfile->FirstParameter()), Precision::Confusion())) {
+                isDuplicate = true;
+                break;
+            }
+        }
+        if (!isDuplicate) {
+            uniqueProfiles.push_back(profile);
+        }
+    }
```
Comment (rainman110) [inline src/internal/InterpolateCurveNetwork.cpp:70]: Not sure whether checking inly the first point is enough.

Theoretically, the other curve could also be reversed etc.

I think it would make sense the factor-out a seperate function

    bool curvesAreSame(curve1, curve2)
    
that you are calling here. This would make it easier to extend the function in case of problems.

File: src/internal/InterpolateCurveNetwork.cpp [hunk 3]
```diff
@@ -45,90 +45,109 @@ InterpolateCurveNetwork::InterpolateCurveNetwork(const std::vector<Handle (Geom_
...
+
+    std::vector<Handle(Geom_Curve)> uniqueGuides;
+    for (const auto& guide : guides) {
+        bool isDuplicate = false;
+        for (const auto& uniqueGuide : uniqueGuides) {
+            if (guide->Value(guide->FirstParameter()).IsEqual(uniqueGuide->Value(uniqueGuide->FirstParameter()), Precision::Confusion())) {
+                isDuplicate = true;
+                break;
+            }
```
Comment (rainman110) [inline src/internal/InterpolateCurveNetwork.cpp:79]: Same here

File: src/internal/InterpolateCurveNetwork.cpp [hunk 4]
```diff
@@ -201,7 +285,7 @@ void InterpolateCurveNetwork::MakeCurvesCompatible()
         newParametersGuides.push_back(sum / nGuides);
     }
 
-    if (newParametersProfiles.front() > 1e-5 || newParametersGuides.front() > 1e-5) {
+    if (newParametersProfiles.front() > m_spatialTol || newParametersGuides.front() > m_spatialTol) {
```
Comment (rainman110) [inline src/internal/InterpolateCurveNetwork.cpp:288]: You cannot use spatialTol here, since spatialTol measures a distance. Here, we are comparing values in the parameter range, that is in [0,1]. This will cause problems for input curves with large spatial size.

File: src/internal/InterpolateCurveNetwork.cpp [hunk 5]
```diff
@@ -238,20 +322,20 @@ void InterpolateCurveNetwork::MakeCurvesCompatible()
         }
 
         // eliminate small inaccuracies at the first knot
-        if (std::abs(oldParametersProfile.front()) < 1e-5) {
+        if (std::abs(oldParametersProfile.front()) < m_spatialTol) {
```
Comment (rainman110) [inline src/internal/InterpolateCurveNetwork.cpp:325]: Same here

File: src/internal/InterpolateCurveNetwork.cpp [hunk 6]
```diff
@@ -238,20 +322,20 @@ void InterpolateCurveNetwork::MakeCurvesCompatible()
...
             oldParametersProfile.front() = 0;
         }
 
-        if (std::abs(newParametersProfiles.front()) < 1e-5) {
+        if (std::abs(newParametersProfiles.front()) < m_spatialTol) {
```
Comment (rainman110) [inline src/internal/InterpolateCurveNetwork.cpp:329]: And here

File: src/internal/InterpolateCurveNetwork.cpp [hunk 7]
```diff
@@ -238,20 +322,20 @@ void InterpolateCurveNetwork::MakeCurvesCompatible()
...
             newParametersProfiles.front() = 0;
         }
 
         // eliminate small inaccuracies at the last knot
-        if (std::abs(oldParametersProfile.back() - 1) < 1e-5) {
+        if (std::abs(oldParametersProfile.back() - 1) < m_spatialTol) {
```
Comment (rainman110) [inline src/internal/InterpolateCurveNetwork.cpp:334]: And here

File: src/internal/InterpolateCurveNetwork.cpp [hunk 8]
```diff
@@ -238,20 +322,20 @@ void InterpolateCurveNetwork::MakeCurvesCompatible()
...
             oldParametersProfile.back() = 1;
         }
 
-        if (std::abs(newParametersProfiles.back() - 1) < 1e-5) {
+        if (std::abs(newParametersProfiles.back() - 1) < m_spatialTol) {
```
Comment (rainman110) [inline src/internal/InterpolateCurveNetwork.cpp:338]: You get it ;)

File: src/internal/InterpolateCurveNetwork.cpp [hunk 9]
```diff
@@ -268,20 +352,20 @@ void InterpolateCurveNetwork::MakeCurvesCompatible()
         }
 
         // eliminate small inaccuracies at the first knot
-        if (std::abs(oldParameterGuide.front()) < 1e-5) {
+        if (std::abs(oldParameterGuide.front()) < m_spatialTol) {
```
Comment (rainman110) [inline src/internal/InterpolateCurveNetwork.cpp:355]: Here

File: src/internal/InterpolateCurveNetwork.cpp [hunk 10]
```diff
@@ -268,20 +352,20 @@ void InterpolateCurveNetwork::MakeCurvesCompatible()
...
             oldParameterGuide.front() = 0;
         }
 
-        if (std::abs(newParametersGuides.front()) < 1e-5) {
+        if (std::abs(newParametersGuides.front()) < m_spatialTol) {
```
Comment (rainman110) [inline src/internal/InterpolateCurveNetwork.cpp:359]: Here as well

File: src/internal/InterpolateCurveNetwork.cpp [hunk 11]
```diff
@@ -268,20 +352,20 @@ void InterpolateCurveNetwork::MakeCurvesCompatible()
...
             newParametersGuides.front() = 0;
         }
 
         // eliminate small inaccuracies at the last knot
-        if (std::abs(oldParameterGuide.back() - 1) < 1e-5) {
+        if (std::abs(oldParameterGuide.back() - 1) < m_spatialTol) {
```
Comment (rainman110) [inline src/internal/InterpolateCurveNetwork.cpp:364]: Here

File: src/internal/InterpolateCurveNetwork.cpp [hunk 12]
```diff
@@ -268,20 +352,20 @@ void InterpolateCurveNetwork::MakeCurvesCompatible()
...
             oldParameterGuide.back() = 1;
         }
 
-        if (std::abs(newParametersGuides.back() - 1) < 1e-5) {
+        if (std::abs(newParametersGuides.back() - 1) < m_spatialTol) {
```
Comment (rainman110) [inline src/internal/InterpolateCurveNetwork.cpp:368]: And here

**Review by rainman110 — CHANGES_REQUESTED** [review]

**Comment by Shkolik** [comment]
> You could add a second function (ideally using overloads) with Geom_BSplineCurve arguments. The "old" function then should convert the Geom_Curves and call the B-Spline-based function.
> 
> Maybe lets do this in a separate PR.

That's exactly what I was thinking of.

File: src/internal/InterpolateCurveNetwork.cpp (hunk 1 above)
Comment (Shkolik) [reply src/internal/InterpolateCurveNetwork.cpp:39]: My bad. VS automatically applies formatting and it's hard not to miss some pieces.

File: src/internal/InterpolateCurveNetwork.cpp (hunk 4 above)
Comment (Shkolik) [reply src/internal/InterpolateCurveNetwork.cpp:288]: I understand when it could became a problem... Ok, lets at least use Precision::Intersection() here and on other places where we compare parameters. It's the same 1e-5, but at least we don't have "magic numbers" scattered in a code.

Comment (rainman110) [reply src/internal/InterpolateCurveNetwork.cpp:288]: > I understand when it could became a problem... Ok, lets at least use Precision::Intersection() here and on other places where we compare parameters. It's the same 1e-5, but at least we don't have "magic numbers" scattered in a code.

Great suggestions!

Comment (Shkolik) [reply src/internal/InterpolateCurveNetwork.cpp]: My bad. It appear that Precision::Intersection() is 1e-9 - probably to much for what we are using it. I changed it to Precision::Approximation() that is 1e-6. Will see how it will behave.

File: src/internal/InterpolateCurveNetwork.cpp (hunk 2 above)
Comment (Shkolik) [reply src/internal/InterpolateCurveNetwork.cpp]: Changed function to check if curves share both ends. It will eliminate but will produce a false error when user try to create fully closed surface. Yes, it's not supported by now, and not sure if it's achievable at all, but still misleading error message. I will think about it little bit more and, if something better came in mind, I'll make separate PR for that.

**Comment by Shkolik** [comment]
I just realized, that if I implement InterpolateCurveNetwork conctructor to take Geom_BSplineCurve instead of Geom_Curve then we don't need separate function to compare curves for duplicated. Geom_BSplineCurve  has IsEqual method and we can use it like duplicate = guide->IsEqual(uniqueGuide, Precision::Confusion()) - another reason to do so.

**Comment by rainman110** [comment]
The helibody test case is now failing:

BSplineAlgorithms/GordonSurface.testFromBRep/8, where GetParam() = "helibody"

I have not yet looked into, why it happens though. An exception is thrown

**Comment by rainman110** [comment]
The tests start failing from commit ec9b8c413399089160ca1d06e9df89f6429ad5ba. In this case multiple tests are failing. 

Your last commit (35fd291036fd365da0c88d61eebb20348a13c0a4) fixes many issues, but the helicopter case is still faulty.

**Comment by rainman110** [comment]
I think, the problems stem from the linear curve reparametrization  right at the start into the interval [0,1]:

https://github.com/rainman110/occ_gordon/blob/2386a5de2ba79e08e61cd6fa1ab1c1c987f1bdff/src/internal/BSplineAlgorithms.cpp#L826C6-L826C45

The function BSplCLib::Reparametrize itself uses some tolerance internally. Lets say the first knot vector is 5e-6, then the function will not set the knot vector to zero but it will keep it there. 

Later on, in the gordon surface builder, these knot will be set to zero based on the tolerance (1e-5). Since 5e-6 is larger than 1e-6 (your new tolerance), the function raises.

**Comment by Shkolik** [comment]
> Later on, in the gordon surface builder, these knot will be set to zero based on the tolerance (1e-5). Since 5e-6 is larger than 1e-6 (your new tolerance), the function raises.

Ok, we know that 1e-5 is working, lets stick with this tolerance then, but I'll define it as a macro.

**Comment by Shkolik** [comment]
Should be fine now

**Review by rainman110 — APPROVED** [review]
@Shkolik Looks good. I did a small code cleanup,  which not change the semantics though.

Nice work. Thank you for your contribution. I think it will definitely improve the usability of the function.

**Comment by Shkolik** [comment]
> @Shkolik Looks good. I did a small code cleanup, which not change the semantics though.
> 
> Nice work. Thank you for your contribution. I think it will definitely improve the usability of the function.

Thank you! Sorry that you had to spend so much time reviewing my amateur code :) Last time I was writing something useful in C++ was 2 decades ago... I'll submit couple more PR later with overload of entry point and, I think, it's a good idea to "un-hardcode" other tolerances and put them in one place. 

BTW, I have a question. I was digging your code and found that you used your own method to interpolate points to spline. It appear that python port using OCC method and it produce slightly different result. Not sure what I like more, but just curious why you decide to reimplement interpolation? 

For example building a surface using test data from nacelle interpolation of first profile intersection points (intake) give following results:

For your method:
```
Degree: 3 / NbPoles: 7
Closed: 1 / Periodic: 0
Knots: ['0.00', '0.28', '0.50', '0.80', '1.00']
Mults: [' 4', ' 1', ' 1', ' 1', ' 4']
```

for OCC interpolate python method:
```
Degree: 3
NbPoles: 5
Knots: 5 ['0.00', '0.28', '0.50', '0.80', '1.00' ]
Mults: [2, 1, 1, 1, 2]
Periodic: True
```

I think some of the problems in a python port might be from incorrect interpolation.

**Comment by rainman110** [comment]
> BTW, I have a question. I was digging your code and found that you used your own method to interpolate points to spline. It appear that python port using OCC method and it produce slightly different result. Not sure what I like more, but just curious why you decide to reimplement interpolation?

I quickly found out, that I had not enough control over the interpolation result. In particular, i needed to precisely steer, at what parameter and also which knot vectors are used by the interpolation. Otherwise, you cannot guaranty the interpolation  property of the gordon surface.

For the curve interpolation (and also for the surface interpolation), i used textbook methods from "the NURBS Book" and "Curves and Surfaces from CAGD". I multiplicity of 4 for the first and last knot is pretty standard for a Degree 3 B-spline interpolation.

I suppose the difference in both methods is, that my result does not return a periodic spline (it is periodic, but the isPeriodic function will tell false). My result is just a clamped spline with periodic properties. It was simply easier to work with normal clamped splines, since I don't really need to distinguish from from the algorithmic point of view, whether the problem is closed or not (except from small parts to achieve smoothness at the junctions).
